cfd_protocol/src/lib.rs

```diff
@@ -558,31 +621,161 @@ mod tests {
             };
 
             let mut tx_refund = self.inner;
-            self.lock_output_descriptor
+            self.commit_descriptor
                 .satisfy(&mut tx_refund.input[0], satisfier)?;
 
             Ok(tx_refund)
         }
     }
 
+    #[derive(Debug, Clone)]
+    struct PunishTransaction {
+        inner: Transaction,
+    }
+
+    impl PunishTransaction {
+        fn new(
+            commit_tx: &CommitTransaction,
+            address: &Address,
+            amount: Amount,
+            encisg: EcdsaAdaptorSignature,
+            sk: SecretKey,
+            (revocation_them_sk, revocation_them_pk): (SecretKey, PublicKey),
+            publishing_them_pk: PublicKey,
+            revoked_commit_tx: Transaction,
+        ) -> Result<Self> {
+            // CommitTransaction has only one input
+            let input = revoked_commit_tx.input[0].clone();
+
+            // Extract all signatures from witness stack
+            let mut sigs = Vec::new();
+            for witness in input.witness.iter() {
+                let witness = witness.as_slice();
+
+                let res = bitcoin::secp256k1::Signature::from_der(&witness[..witness.len() - 1]);
+                match res {
+                    Ok(sig) => sigs.push(sig),
+                    Err(_) => {
+                        continue;
+                    }
+                }
+            }
```

Comment:
Such kind of stuff can be nicely extracted into extension traits so the actual flow of this function is less noisy and composed of more powerful low-level functionality.

Reply:
Interesting idea. I rarely consider using the extension trait pattern, so it's very useful to get feedback on where they can be used!

In this case, would your idea be to extend TxIn with (e.g.) TxInExt and define a extract_signatures trait method? I guess in this particular instance since this all remains private API of the library we could just extract a free function too.

Reply:
Yeah, the only advantage of an extension trait over a free function is that it is easier to discover because the IDE will give you auto-completion for it.

In this case, yes probably put it on `TxIn`.

Reply:
We haven't done this yet because this it's just refactoring, but I would like to try this out: https://github.com/comit-network/hermes/issues/20. 

I could have probably just done it, but we can use that issue to remember that there might be other opportunities to use extension traits.